sdk/lib/cmlib/cmcheck.c

```diff
@@ -0,0 +1,1800 @@
+                    return CmStatusCode;
+                }
+
+                /*
+                 * We have repaired the hive. Since the damaged
+                 * cell has been removed, we must restart the
+                 * loop and start over at the beginning.
+                 */
+                DPRINT1("CmpValidateRegistryInternal(): Hive repaired, restarting the validation loop...\n");
+                goto RestartValidation;
+            }
+
+            /*
+             * The key is in perfect shape. If we have advanced
+             * the stack depth then check the lexicographical
+             * order of the keys as well.
+             */
+            if (StackDepth > 0 &&
+                CM_CHECK_REGISTRY_SUCCESS(CmStatusCode))
+            {
+                if (WorkState[StackDepth - CMP_PRIOR_STACK].Sibling != HCELL_NIL)
+                {
+                    if (!CmpValidateLexicalOrder(Hive,
+                                                 CurrentCell,
+                                                 WorkState[StackDepth - CMP_PRIOR_STACK].Sibling))
+                    {
+                        /*
+                         * The lexicographical order is bad,
+                         * attempt to heal the hive.
+                         */
+                        if (!CmpRepairParentKey(Hive, CurrentCell, ParentCell))
+                        {
+                            /*
+                             * If we reach here then either self healing
+                             * is not possible or we couldn't repair the
+                             * parent list from damage. Bail out.
+                             */
+                            DPRINT1("CmpValidateRegistryInternal(): The lexicographical order is invalid (sibling %lu, current cell %lu)!\n",
+                            CurrentCell, WorkState[StackDepth - CMP_PRIOR_STACK].Sibling);
+                            CmpFree(WorkState, WorkStateLength);
+                            return CM_CHECK_REGISTRY_BAD_LEXICOGRAPHICAL_ORDER;
+                        }
+
+                        /*
+                         * We have repaired the hive. Since the damaged
+                         * cell has been removed, we must restart the
+                         * loop and start over at the beginning.
+                         */
+                        DPRINT1("CmpValidateRegistryInternal(): Hive repaired, restarting the validation loop...\n");
+                        goto RestartValidation;
+                    }
+                }
+
+                /* Assign the prior sibling for upcoming iteration */
+                WorkState[StackDepth - CMP_PRIOR_STACK].Sibling = CurrentCell;
+            }
+        }
+
+        /* Obtain a node for this key */
+        KeyNode = (PCM_KEY_NODE)HvGetCell(Hive, CurrentCell);
+        if (!KeyNode)
+        {
+            /* We couldn't be able to get the node, bail out */
```

Comment:
superfluous comment. Try to comment less! Focus on the "why" and not the "what". Especially when the "what" is obvious.